### PR TITLE
Laboratory work #4, Alexander Babii - 19FPL1

### PR DESCRIPTION
**Pull request by asbabiy**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
Deadline is coming

File: lab_4/likelihood_text_generator_score_eight_test.py [hunk 1]
```diff
@@ -157,7 +157,8 @@ def test_generate_next_word_incorrect_context(self):
 
         trie = NGramTrie(3, encoded)
 
-        bad_inputs = [[], {}, (2000, 1000, ), None, 9, 9.34, True]
+        # I was told that it (2000, 1000,) can be changed to (2000,)
```
Comment (ValeriyaKuznetsova) [inline lab_4/likelihood_text_generator_score_eight_test.py:160]: I don't know about that :) This test is valid (these ids cannot exist with the present text)

File: lab_4/main.py [hunk 2]
```diff
@@ -1,74 +1,232 @@
...
-        pass
+        if (not isinstance(context, tuple) or
+                len(context) + 1 != self._n_gram_trie.size or
+                context[0] > len(self._word_storage.storage)):
+            raise ValueError
+
+        max_size = len(context) + 1
+        self._n_gram_tries = [trie for trie in self._n_gram_tries if trie.size <= max_size]
+
+        for trie in self._n_gram_tries:
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: are we going from max size to min here?

File: lab_4/main.py [hunk 3]
```diff
@@ -1,74 +1,232 @@
...
+            self._n_gram_trie = trie
+
+            if n_gram := super()._generate_next_word(context):
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: what is this `n_gram`? it isn't mentioned before...

File: lab_4/main.py [hunk 4]
```diff
@@ -1,74 +1,232 @@
...
+                return n_gram
+            context = context[:-1]
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: why :-1?

File: lab_4/main.py [hunk 5]
```diff
@@ -1,74 +1,232 @@
...
+    if not isinstance(path_to_saved_model, str):
+        raise ValueError
+
+    with open(path_to_saved_model) as obj:
+        attrs = dict(ast.literal_eval(obj.read()))
+
+    storage = type('WordStorage', (), attrs['_word_storage'])
+    trie = type('NGramTrie', (), attrs['_n_gram_trie'])
+
+    class Model(NGramTextGenerator):
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: why do we need to create class here?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
have some questions

File: lab_4/main.py (hunk 5 above)
Comment (asbabiy) [reply lab_4/main.py]: Well, we actually don't, I just forgot to remove these lines after changing the serializiation strategy)

File: lab_4/main.py (hunk 2 above)
Comment (asbabiy) [reply lab_4/main.py]: Well, not always. `_n_gram_tries` is a list with `NGramTrie` objects. It is not sorted, so at first 5-grams can be tested, then 3-grams, then 4-grams and so on. I just use list comprehension to choose those ones which size isn't higher than max_size.

File: lab_4/main.py (hunk 3 above)
Comment (asbabiy) [reply lab_4/main.py]: `n_gram` is declared and implicitly converted to the `bool` type right in the `if ...` statement using the walrus operator. It is a result of the `super()._generate_next_word(context)` so it is just an encoded generated word.

File: lab_4/main.py (hunk 2 above)
Comment (asbabiy) [reply lab_4/main.py:185]: Made it sorted

Comment (asbabiy) [reply lab_4/main.py]: Made it sorted.

File: lab_4/main.py (hunk 4 above)
Comment (asbabiy) [reply lab_4/main.py]: Changed implementation a bit, removed it

File: lab_4/main.py (hunk 3 above)
Comment (asbabiy) [reply lab_4/main.py]: And removed this too

**Pull request merged**